cmd/socat/main.go

```diff
@@ -0,0 +1,143 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"strings"
+
+	"github.com/spf13/cobra"
+)
+
+func main() {
+	cmd := newRootCmd()
+	err := cmd.Execute()
+	if err != nil {
+		os.Exit(1)
+	}
+}
+
+func newRootCmd() *cobra.Command {
+	var uniDir bool
+	cmd := cobra.Command{
+		Use:   "socat <address> <address>",
+		Short: "Minimalistic socat.",
+		Long: `Minimalistic socat.
+Implements only TCP, OPEN and stdio ("-") addresses with no options.
+Only supported flag is -u.`,
+		Args: cobra.ExactArgs(2),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			right, err := createConnection(args[0])
+			if err != nil {
+				return err
+			}
+			left, err := createConnection(args[1])
+			if err != nil {
+				return err
+			}
+			return connect(right, left, uniDir)
+		},
+	}
+
+	cmd.Flags().BoolVarP(&uniDir, "", "u", false, "unidirectional mode (left to right)")
+
+	return &cmd
+}
+
+func createConnection(address string) (connection, error) {
+	if address == "-" {
+		return stdio, nil
+	}
+	parts := strings.SplitN(address, ":", 2)
+	if len(parts) != 2 {
+		return nil, fmt.Errorf("cannot parse address: %q", address)
+	}
+	typ := strings.ToLower(parts[0])
+	parts = strings.Split(parts[1], ",")
+	if len(parts) > 1 {
+		_, _ = fmt.Fprintf(os.Stderr, "flags ignored: %q\n", parts[1])
+	}
+	addr := parts[0]
+	switch typ {
+	case "tcp", "tcp4", "tcp6":
+		var laddr net.TCPAddr
+		raddr, err := net.ResolveTCPAddr(typ, addr)
+		if err != nil {
+			return nil, fmt.Errorf("name does not resolve: %w", err)
+		}
+		return net.DialTCP(typ, &laddr, raddr)
+	case "open":
+		return os.OpenFile(addr, os.O_RDWR, 0644)
+	}
+	return nil, fmt.Errorf("unsupported address: %q", address)
```

Comment:
Might be nice to stick this in the `default` case of the switch statement and say `unsupported protocol`